### PR TITLE
Integration test infrastructure

### PR DESCRIPTION
**Pull request by kotfic**
Still planning on writing some code that wraps the command and allows us to do fully automated testing.  To check out just the command you can run:

``` bash
    cumulus full_status > before
    cumulus -v create_profile \
           create_cluster \
           launch_cluster \
           create_volume \
           attach_volume

    cumulus full_status > during

    cumulus -v detach_volume \
           delete_volume \
           terminate_cluster \
           delete_cluster \
           delete_profile

    cumulus full_status > after
```

Then, to verify the state you should then be able to run something like:

```
sdiff -w 300 before during
sdiff -w 300 during after
sdiff -w 300 before after
```

### CODE REVIEW

File: cumulus/ansible/tasks/volume.py [hunk 1]
```diff
@@ -38,16 +38,19 @@ def attach_volume(profile, cluster, instance, volume, path,
...
         'volume_name': volume['name'],
-        'volume_size': volume['size'],
         'volume_zone': volume['zone'],
         'instance_id': instance['instance_id'],
         'path': path,
         'ansible_ssh_private_key_file': _key_path(profile),
         'ansible_user': cluster['config']['ssh']['user']
     }
 
+    if volume['ec2']['id']:
```
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: Is 'ec2.id. always part of a cluster? We might way the use jsonpath the extract out this value.

File: cumulus/scripts/command.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+def list_profiles(proxy):
+    print("Listing profiles:")
+
+    keys    = [key('name'), key('status'), key('_id'),
+               key('regionName'), key('cloudProvider')]
+    headers = ['Name', 'Status', 'Profile ID',  'Region', 'Cloud Provider']
+
+    print tabulate([[f(p) for f in keys] for p in proxy.profiles],
+                   headers=headers)
+    print "\n"
```
Comment (cjh1) [inline cumulus/scripts/command.py]: Can we use print(...)

File: cumulus/taskflow/core/test/mytaskflows.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def delete(self, *args, **kwargs):
 
 @command.task
 def regular_task():
-    print 'regular_task!!'
+    'regular_task!!'
```
Comment (cjh1) [inline cumulus/taskflow/core/test/mytaskflows.py]: What is this change for?

File: cumulus/transport/ssh.py [hunk 4]
```diff
@@ -53,6 +53,13 @@ def __enter__(self):
 
         username = parse('config.ssh.user').find(self._cluster)[0].value
         hostname = parse('config.host').find(self._cluster)[0].value
+
+        port = parse('config.port').find(self._cluster)
```
Comment (cjh1) [inline cumulus/transport/ssh.py]: This should probably be 'config.ssh.port'?

File: tests/integration/integration.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+    _, stdout, stderr = client.exec_command('echo "%s" >> ~/.ssh/authorized_keys' % key)
+    assert bool(stdout.read()) == False
+    assert bool(stderr.read()) == False
+
+    proxy.put('clusters/%s/start' % proxy.cluster['_id'])
+    proxy.wait_for_status('clusters/%s/status' % proxy.cluster['_id'],
+                          'running', timeout=60)
+
+
+    import pudb; pu.db
```
Comment (cjh1) [inline tests/integration/integration.py]: Debug?

File: tests/integration/integration.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    r = proxy.client.listItem(output_folder_id)
+    assert len(r) == 4
+
+
+    # Clean up
+    ctx.invoke(delete_cluster, cluster_section=cluster_section)
+    ctx.invoke(delete_profile, profile_section=profile_section)
+
+
+
```
Comment (cjh1) [inline tests/integration/integration.py]: I wonder the different tests should be separated out into a couple of files and then import into the main test driver code.

File: requirements-dev.txt [hunk 7]
```diff
@@ -2,3 +2,5 @@ mock==1.0.1
 httmock==1.2.2
 flake8-quotes==0.2.4
 Flask==0.10.1
+click==6.6
+tabulate==0.7.5
```
Comment (cjh1) [inline requirements-dev.txt]: I we are added an entry point for the command line tool, these should probably be in requirements.txt?

**Comment by kotfic** [comment]
A few things to change after meeting with @cjh1 

+ [ ] Reorganizing [wait_for_status](https://github.com/Kitware/cumulus/blob/integration_test_infrastructure/cumulus/scripts/utils.py#L534) into [check_log](https://github.com/Kitware/cumulus/blob/integration_test_infrastructure/cumulus/scripts/utils.py#L493-L516) because of complexity linting has introduced a bug.  Not all ansible log messages are displaying (task still completes correctly). 
+ [ ] CLI tool should chain commands in a more natural way (e.g.  instead of ```cumulus create_cluster```  it should be ```cumulus cluster create```,  ```cumulus cluster launch``` etc)
+ [ ] CLI tools should not create resources just to delete them (e.g. ```cumulus delete_cluster``` this is fine for testing)
+ [ ] Test infrastructure should warn (instead of fail)  if profiles/clusters/etc  exist in the database
+ [ ] Add click documentation to subcommands
+ [ ] Add documentation for CLI command
+ [ ] Add documentation for integration testing command

File: cumulus/transport/ssh.py (hunk 4 above)
Comment (kotfic) [inline cumulus/transport/ssh.py]: I was thinking port should be with host,  so should host be in config.ssh also?

**Comment by kotfic** [comment]
@cjh1 This is ready for review.  You may want to start with documentation for the [command](https://github.com/Kitware/cumulus/tree/integration_test_infrastructure/cumulus/scripts/) and the [integration tests](https://github.com/Kitware/cumulus/blob/integration_test_infrastructure/tests/integration/)

I have not tested that all the integration tests are passing (e.g. test_traditional).  As long as the ansible related tests are working for you,  i'd like to fix those in another PR (one that also removed the 'legacy' integration scripts in tests/integration/).

**Comment by cjh1** [comment]
@kotfic Thanks for pushing this forward, this is going to be very useful. I will review soon.

**Comment by cjh1** [comment]
@kotfic I am starting to look at this, I think it currently making the assumption that all cluster are cloud clusters. In my current environment I have several traditional clusters. If I issue:

```
cumulus list_clusters
```

I get:

```
Traceback (most recent call last):
  File "/home/cjh/virtualenv/cumulus/bin/cumulus", line 9, in <module>
    load_entry_point('cumulus==0.9.0', 'console_scripts', 'cumulus')()
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 1086, in invoke
    rv.append(sub_ctx.command.invoke(sub_ctx))
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/home/cjh/virtualenv/cumulus/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/cjh/work/source/cumulus/cumulus/scripts/command.py", line 81, in list_clusters
    print tabulate([[f(c) for f in keys] for c in proxy.clusters],
  File "/home/cjh/work/source/cumulus/cumulus/scripts/utils.py", line 62, in _profile
    return profile_dict[instance['profileId']]
KeyError: 'profileId'
```

File: cumulus/ansible/tasks/volume.py (hunk 1 above)
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: I think the comment above is still valid.

File: cumulus/scripts/command.py [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+
+
+###############################################################################
+#   Profile Commands
+#
+
+@cli.group()
+@click.option('--profile_section', default='profile',
+              help=PROFILE_SECTION_HELP)
+@pass_proxy
```
Comment (cjh1) [inline cumulus/scripts/command.py]: How does this decorator know how to extract the correct instance?

File: cumulus/taskflow/core/test/mytaskflows.py (hunk 3 above)
Comment (cjh1) [inline cumulus/taskflow/core/test/mytaskflows.py]: ?

File: cumulus/transport/ssh.py (hunk 4 above)
Comment (cjh1) [inline cumulus/transport/ssh.py]: port is specific to ssh where as host is not.

File: cumulus/scripts/command.py (hunk 8 above)
Comment (kotfic) [inline cumulus/scripts/command.py]: @cjh1 not quite sure what you're asking?

Comment (cjh1) [inline cumulus/scripts/command.py]: @kotfic Not not sure either :smile: !

**Comment by cjh1** [comment]
@kotfic Here is some JSON to create a trad cluster: 

```json
 body = {
            'config': {
                'ssh': {
                    'user': 'billy'
                },
                'host': 'hell'
            },
            'name': 'TradIntegrationTest',
            'type': 'trad'
        }
```

**Comment by kotfic** [comment]
@cjh1 this should fix the traditional cluster issue. thanks!

**Pull request merged**